org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/concurrent/RequestManager.xtend

```diff
@@ -7,111 +7,134 @@
-				try {
-					r.lock
-					toCancel += cancelIndicator
-					cancelIndicator.checkCanceled
-					result.complete(request.apply(cancelIndicator))
-				} catch (Throwable e) {
-					result.completeExceptionally(e)
-				} finally {
-					toCancel -= cancelIndicator
-					r.unlock()
-				}
-			]
-		]
-		return result
+	def <V> CompletableFuture<V> runRead((CancelIndicator)=>V cancellable) {
+		return submit(
+			new ReadRequest(cancellable, parallel)
+		)
+	}
+
+	def <U, V> CompletableFuture<V> runWrite(()=>U nonCancellable, (CancelIndicator, U)=>V cancellable) {
+		val cancelFuture = cancel()
+		return submit(
+			new WriteRequest(nonCancellable, cancellable, cancelFuture)
+		)
+	}
+
+	protected def <V> CompletableFuture<V> submit(AbstractRequest<V> request) {
+		requests += request
+		queue.submit(request)
+		return request.get
+	}
+
+	protected def CompletableFuture<Void> cancel() {
+		val localRequests = requests
+		requests = newArrayList
+
+		val cfs = newArrayList
+		for (request : localRequests) {
+			request.cancel
+			cfs += request.get
+		}
+		return CompletableFuture.allOf(cfs)
+	}
+
+	protected def boolean isCancelException(Throwable t) {
+		if(t === null) return false;
+		val cause = if(t instanceof CompletionException) t.cause else t
+		return operationCanceledManager.isOperationCanceledException(cause);
 	}
-	
-	var CompletableFuture<Void> writePending = CompletableFuture.completedFuture(null)
-
-	def <U, V> CompletableFuture<V> runWrite(()=>U nonCancellable, (CancelIndicator, U)=>V request) {
-		val result = new CompletableFuture<V>()
-		val localWritePending = new CompletableFuture<Void>()
-		this.writePending = localWritePending
-		parallel.submit[
-			val cancelIndicator = new RequestCancelIndicator(result)
+
+}
+
+@FinalFieldsConstructor
+class ReadRequest<V> extends AbstractRequest<V> {
+
+	val (CancelIndicator)=>V cancellable
+	val ExecutorService executor
+
+	override void run() {
+		if(result.cancelled) return;
+		executor.submit [
 			try {
-				cancel()
-				w.lock
-				localWritePending.complete(null)
-				// run the non-cancelable work
-				val intermediateResult = nonCancellable.apply()
-				
-				// run the cancelable work
-				toCancel += cancelIndicator
 				cancelIndicator.checkCanceled
-				result.complete(request.apply(cancelIndicator, intermediateResult))
-			} catch (Throwable e) {
+				result.complete(cancellable.apply(cancelIndicator))
+			} catch(Throwable e) {
 				result.completeExceptionally(e)
-			} finally {
-				toCancel -= cancelIndicator
-				w.unlock
 			}
 		]
-		return result
 	}
 
-	protected def void cancel() {
-		for (cancellable : toCancel) {
-			cancellable.cancel
-			toCancel -= cancellable
+}
+
+@FinalFieldsConstructor
+class WriteRequest<U, V> extends AbstractRequest<V> {
+
+	val ()=>U nonCancellable
+	val (CancelIndicator, U)=>V cancellable
+	val CompletableFuture<Void> previous
+
+	override void run() {
+		try {
+			previous.join
+		} catch(Throwable e) {
+			/*no-op */
```

Comment:
Can you clarify why it's ok to forget about the exceptional outcome of the previous future?

Reply:
We are not interested in results, only to make sure that all previous requests are finished before running next write request. Should I add a comment?

Reply:
Yes, that would help to understand the reasoning behind `/* no-op */`